### ISSUE TITLE
Could customConditions exports be allowed to come before `types` in `exports`?

### ISSUE THREAD

**Issue by tylersayshi** [publint/publint#176]
for this rule: https://publint.dev/rules#exports_types_should_be_first

I'd like to be be be able to use `customConditions` as the first export without a lint error for types being first. Maybe `.{ts,tsx}` exports could be allowed to come before `"types"`?

The approach for this is described in this blog from colinhacks: https://colinhacks.com/essays/live-types-typescript-monorepo in the subsection "5. Custom conditions in "exports"".

**Comment by bluwy** [comment]
I think that makes sense 👍 Though I think we would also have to disable file existence check for `.ts` and `.tsx` as they can be chosen to not be published later.

**Comment by tylersayshi** [comment]
Makes sense! I can make a PR for this if you'd like?

**Comment by bluwy** [comment]
No worries. I've already started working on this locally 😅

**Comment by tylersayshi** [comment]
I think some of the assumptions here might have been flawed. The custom condition I was adding should have been one level higher, instead of inside `import` and `require`. Here is the fix for the initial PR I mentioned this from: https://github.com/TanStack/query/pull/8959

I think this may change how the publint rule should work, but either way, I just wanted to mention to see what you think.

**Comment by bluwy** [comment]
Is there currently false warnings/errors with the change? But I think yes that the logic currently should be tracking across objects while crawling, rather than the current object only. I actually implemented that halfway initially but thought it didn't need that. Though IIUC the logic should still indirectly not misflag the changed setup.

**Comment by tylersayshi** [comment]
> Is there currently false warnings/errors with the change?

no, my worry is that I suggested adding support inside the `import` & `require` objects for ignoring customConditions, but the custom condition should have existed outside the import/require objects to begin with. [b34ea94](https://github.com/publint/publint/commit/b34ea9433fb7457ac7f5f8ade972e1589ca9c89e) added support to check inside the objects and consider those valid if I'm not mistaken.

There's no mistake warnings or errors, but I think the commit added for this issue is checking to ignore more than is needed to be checked. Maybe that does not matter, but it affectively seems like dead code when this pattern is done correctly.

**Comment by bluwy** [comment]
I think it should still work fine with the commit. The checks are done on any objects that has the `"types"` condition, so in your case with the custom condition moved out, the `"types"` isn't there and it indirectly worked. I think it's fine for now as the checks are always conservative first.

It could be improved better though to handle cases like this:

```json
{
  "exports": {
    "browser": "./browser.js",
    "import": {
      "types": "./index.d.ts",
      "default": "./index.js"
    },
    "require": {
      "types": "./index.d.xts",
      "default": "./index.xjs"
    }
  }
}
```

Because looking it as a whole, the `"types"` won't be effective if the browser condition matched, so publint should flag it (but doesn't without extra work now). But if it does work, then it also needs to account for the custom conditions (or checking exporting plain .ts) case to not flag issues with this.
